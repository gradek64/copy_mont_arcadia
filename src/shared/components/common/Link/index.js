import React from 'react'
import PropTypes from 'prop-types'
import { Link as RRLink } from 'react-router'

/* eslint-disable jsx-a11y/anchor-has-content */
export default function Link({
  isExternal = false,
  to,
  onlyActiveOnIndex = false,
  ...restProps
}) {
  if (isExternal) return <a {...restProps} href={to} />

  return <RRLink {...restProps} to={to} onlyActiveOnIndex={onlyActiveOnIndex} />
}

Link.propTypes = {
  ...RRLink.propTypes,
  children: PropTypes.node.isRequired,
  isExternal: PropTypes.bool,
  onlyActiveOnIndex: PropTypes.bool,
}
