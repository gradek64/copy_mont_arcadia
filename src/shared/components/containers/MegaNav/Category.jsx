import React, { memo } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { isEmpty, equals } from 'ramda'

import { heDecode, replaceUrl } from '../../../lib/html-entities'

import SubNav from './SubNav'
import Link from '../../common/Link'

const getUrl = ({
  apiEnvironment,
  category,
  isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled,
}) => {
  const { redirectionUrl, url } = category
  const toUrl =
    !isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled && redirectionUrl
      ? redirectionUrl
      : url
  return apiEnvironment === 'prod' ? toUrl : replaceUrl(toUrl)
}

const Category = ({
  category,
  onHoverCategory,
  unselectCategory,
  apiEnvironment,
  isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled,
  touchEnabled,
  megaNavSelectedCategory,
}) => {
  const { categoryId, label, columns } = category
  const categoryClasses = classnames('MegaNav-category', {
    [`category_${categoryId}`]: categoryId,
    'MegaNav-category--isNotTouch': !touchEnabled,
    'MegaNav-category--isTouch': touchEnabled,
  })

  const title = heDecode(label)
  const categoryUrl = getUrl({
    apiEnvironment,
    category,
    isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled,
  })

  // Comment from matol, from March 15th, 2019 2:22pm - DES-5446 handle /blog in mega nav main level links (#8053)
  // The following is only temporary. A more generic/extendable solution
  // will be implemented as part of https://arcadiagroup.atlassian.net/browse/DES-5457
  const urlToBeServerSideRendered = ['/blog/']
  const linkIsExternal = urlToBeServerSideRendered.includes(categoryUrl)
  const shouldShowSpan = touchEnabled && !isEmpty(columns)
  const linkCommonProperties = {
    className: 'MegaNav-categoryLink',
    title,
    onTouchStart: () => onHoverCategory({ categoryId, categoryUrl }),
  }
  return (
    <li
      key={categoryId}
      className={categoryClasses}
      onTouchStart={() => onHoverCategory({ categoryId })}
      onMouseEnter={() => onHoverCategory({ categoryId })}
    >
      {shouldShowSpan ? (
        <span {...linkCommonProperties} style={{ cursor: 'pointer' }}>
          {title}
        </span>
      ) : (
        <Link
          {...linkCommonProperties}
          isExternal={linkIsExternal}
          to={categoryUrl}
          onClick={linkIsExternal ? null : unselectCategory}
        >
          {title}
        </Link>
      )}
      {!isEmpty(columns) && (
        <div
          className={classnames('MegaNav-subNavWrapper', {
            'MegaNav-subNavWrapper--isNotTouch': !touchEnabled,
          })}
        >
          <SubNav
            key={categoryId}
            isActive={categoryId === megaNavSelectedCategory}
            category={category}
            unselectCategory={unselectCategory}
          />
        </div>
      )}
    </li>
  )
}

Category.propTypes = {
  category: PropTypes.object.isRequired,
  onHoverCategory: PropTypes.func.isRequired,
  unselectCategory: PropTypes.func.isRequired,
  apiEnvironment: PropTypes.string,
  isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled:
    PropTypes.bool.isRequired,
  touchEnabled: PropTypes.bool.isRequired,
}

Category.defaultProps = {
  apiEnvironment: '',
}

const areEqual = (prevProps, nextProps) => equals(prevProps, nextProps)

export default memo(Category, areEqual)
