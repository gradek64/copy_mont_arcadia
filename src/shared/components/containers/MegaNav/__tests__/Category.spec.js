import {
  buildComponentRender,
  shallowRender,
} from '../../../../../../test/unit/helpers/test-component'

import Category from '../Category'
import Link from '../../../common/Link'

const requiredProps = {
  category: {
    categoryId: '12345',
    url: 'http://www.topman.com',
    redirectionUrl:
      '/en/tsuk/category/new-in-this-week-2169932/new-in-fashion-6367514/N-8d7Zdgl?No=0&Nrpp=20&siteId=%2F12556',
    columns: [
      {
        span: '0',
        subcategories: {
          0: {
            newIn: [
              { image: { span: 'footer' } },
              { label: 'something to buy' },
            ],
          },
        },
      },
      {
        span: '1',
        subcategories: {
          0: {
            jeans: [{ image: { span: 'footer' } }, { label: 'blue jeans' }],
            0: [{ image: { span: 1 } }, { label: 'red jeans' }],
          },
        },
      },
      {
        span: '2',
        subcategories: {
          0: {
            socks: [{ label: 'winter socks' }],
            jeans: [{ image: { url: '' } }],
          },
        },
      },
    ],
  },
  apiEnvironment: 'prod',
  isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled: true,
  onHoverCategory: jest.fn(),
  unselectCategory: jest.fn(),
  touchEnabled: false,
}

const renderComponent = buildComponentRender(shallowRender, Category)

describe('<Category />', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('@renders', () => {
    beforeEach(() => jest.clearAllMocks())

    it('should not render links for touch devices', () => {
      const props = {
        ...requiredProps,
        touchEnabled: true,
      }
      expect(renderComponent(props).getTree()).toMatchSnapshot()
    })

    it('should render links for touch devices if sub category columns is empty', () => {
      expect(
        renderComponent({
          ...requiredProps,
          category: { columns: [], label: 'category' },
          touchEnabled: true,
        }).getTree()
      ).toMatchSnapshot()
    })

    it('should not render sub category if columns is empty', () => {
      expect(
        renderComponent({
          ...requiredProps,
          category: { columns: [], label: 'category' },
        }).getTree()
      ).toMatchSnapshot()
    })

    it('should render correct state and css classes when touchDetection is true', () => {
      const props = {
        ...requiredProps,
        touchEnabled: true,
      }
      const { wrapper } = renderComponent(props)
      expect(
        wrapper
          .find('li')
          .at(0)
          .hasClass('MegaNav-category--isTouch')
      ).toBe(true)
    })

    it('should render correct state and css classes when touchDetection is false', () => {
      const { wrapper } = renderComponent(requiredProps)
      expect(
        wrapper
          .find('li')
          .at(0)
          .hasClass('MegaNav-category--isNotTouch')
      ).toBe(true)
    })

    it('should render without a className categoryId', () => {
      expect(
        renderComponent({ ...requiredProps, categoryId: '' }).getTree()
      ).toMatchSnapshot()
    })

    it('should render for touch device with no sub categories', () => {
      const { getTree } = renderComponent({
        ...requiredProps,
        category: {
          categoryId: '12345',
          columns: [],
        },
      })
      expect(getTree()).toMatchSnapshot()
    })

    it('should render hyperlink if url associated to the category is /blog/', () => {
      const blogCategoryProps = {
        category: {
          url: '/blog/',
          columns: [],
        },
      }
      const { wrapper } = renderComponent(blogCategoryProps)

      const categoryLink = wrapper.find(Link)
      expect(categoryLink.length).toBe(1)
      expect(categoryLink.props().to).toBe('/blog/')
      expect(categoryLink.props().isExternal).toBe(true)
    })

    it('should render Link if url associated to the category is not blog /blog/', () => {
      const blogCategoryProps = {
        category: {
          url: '/notBlog',
          columns: [],
        },
      }
      const { wrapper } = renderComponent(blogCategoryProps)

      expect(wrapper.find('Link').length).toBe(1)
      expect(wrapper.find('Link').props().to).toBe('/notBlog')
    })
  })

  describe('@Methods', () => {
    describe('getUrl()', () => {
      beforeEach(() => jest.clearAllMocks())

      it('should return the redirectionUrl value if prop isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled is false', () => {
        const { wrapper } = renderComponent({
          ...requiredProps,
          apiEnvironment: 'prod',
          isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled: false,
        })

        const redirectionUrl = requiredProps.category.redirectionUrl
        const toLink = wrapper.find(Link).props().to

        expect(toLink).toEqual(redirectionUrl)
      })

      it('should return the redirectionUrl value if prop isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled is not present', () => {
        const props = {
          ...requiredProps,
        }
        delete props.isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled

        const { wrapper } = renderComponent({
          ...props,
          apiEnvironment: 'prod',
          isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled: false,
        })

        const redirectionUrl = requiredProps.category.redirectionUrl
        const toLink = wrapper.find(Link).props().to

        expect(toLink).toEqual(redirectionUrl)
      })

      it('should return the url value if prop isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled is true', () => {
        const { wrapper } = renderComponent({
          ...requiredProps,
          apiEnvironment: 'prod',
          isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled: true,
        })

        const url = requiredProps.category.url
        const toLink = wrapper.find(Link).props().to

        expect(toLink).toEqual(url)
      })
    })
  })

  describe('@Events', () => {
    describe('mouseenter', () => {
      beforeEach(() => {
        jest.clearAllMocks()
      })

      it('calls onHoverCategory onMouseEnter', () => {
        const { wrapper } = renderComponent(requiredProps)

        expect(requiredProps.onHoverCategory).toHaveBeenCalledTimes(0)
        wrapper.find('li').simulate('mouseenter')
        expect(requiredProps.onHoverCategory).toHaveBeenCalledTimes(1)
      })

      it('calls onHoverCategory ', () => {
        const props = {
          ...requiredProps,
          isFeatureMeganavTopLevelLinksNoRedirectionUrlEnabled: false,
          touchEnabled: true,
        }
        const { wrapper } = renderComponent(props)

        expect(requiredProps.onHoverCategory).toHaveBeenCalledTimes(0)
        wrapper.find('span').simulate('touchstart')
        expect(requiredProps.onHoverCategory).toHaveBeenCalledTimes(1)
        expect(requiredProps.onHoverCategory).toHaveBeenCalledWith({
          categoryId: '12345',
          categoryUrl:
            '/en/tsuk/category/new-in-this-week-2169932/new-in-fashion-6367514/N-8d7Zdgl?No=0&Nrpp=20&siteId=%2F12556',
        })
      })
    })
  })
})
