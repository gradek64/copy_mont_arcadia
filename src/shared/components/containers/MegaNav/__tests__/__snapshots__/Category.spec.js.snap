// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Category /> @renders should not render links for touch devices 1`] = `
<li
  className="MegaNav-category category_12345 MegaNav-category--isTouch"
  key="12345"
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <span
    className="MegaNav-categoryLink"
    onTouchStart={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  />
  <div
    className="MegaNav-subNavWrapper"
  >
    <SubNav
      category={
        Object {
          "categoryId": "12345",
          "columns": Array [
            Object {
              "span": "0",
              "subcategories": Object {
                "0": Object {
                  "newIn": Array [
                    Object {
                      "image": Object {
                        "span": "footer",
                      },
                    },
                    Object {
                      "label": "something to buy",
                    },
                  ],
                },
              },
            },
            Object {
              "span": "1",
              "subcategories": Object {
                "0": Object {
                  "0": Array [
                    Object {
                      "image": Object {
                        "span": 1,
                      },
                    },
                    Object {
                      "label": "red jeans",
                    },
                  ],
                  "jeans": Array [
                    Object {
                      "image": Object {
                        "span": "footer",
                      },
                    },
                    Object {
                      "label": "blue jeans",
                    },
                  ],
                },
              },
            },
            Object {
              "span": "2",
              "subcategories": Object {
                "0": Object {
                  "jeans": Array [
                    Object {
                      "image": Object {
                        "url": "",
                      },
                    },
                  ],
                  "socks": Array [
                    Object {
                      "label": "winter socks",
                    },
                  ],
                },
              },
            },
          ],
          "redirectionUrl": "/en/tsuk/category/new-in-this-week-2169932/new-in-fashion-6367514/N-8d7Zdgl?No=0&Nrpp=20&siteId=%2F12556",
          "url": "http://www.topman.com",
        }
      }
      isActive={false}
      key="12345"
      unselectCategory={[MockFunction]}
    />
  </div>
</li>
`;

exports[`<Category /> @renders should not render sub category if columns is empty 1`] = `
<li
  className="MegaNav-category MegaNav-category--isNotTouch"
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <Link
    className="MegaNav-categoryLink"
    isExternal={false}
    onClick={[MockFunction]}
    onTouchStart={[Function]}
    title="category"
  >
    category
  </Link>
</li>
`;

exports[`<Category /> @renders should render for touch device with no sub categories 1`] = `
<li
  className="MegaNav-category category_12345 MegaNav-category--isNotTouch"
  key="12345"
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <Link
    className="MegaNav-categoryLink"
    isExternal={false}
    onClick={[MockFunction]}
    onTouchStart={[Function]}
  />
</li>
`;

exports[`<Category /> @renders should render links for touch devices if sub category columns is empty 1`] = `
<li
  className="MegaNav-category MegaNav-category--isTouch"
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <Link
    className="MegaNav-categoryLink"
    isExternal={false}
    onClick={[MockFunction]}
    onTouchStart={[Function]}
    title="category"
  >
    category
  </Link>
</li>
`;

exports[`<Category /> @renders should render without a className categoryId 1`] = `
<li
  className="MegaNav-category category_12345 MegaNav-category--isNotTouch"
  key="12345"
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <Link
    className="MegaNav-categoryLink"
    isExternal={false}
    onClick={[MockFunction]}
    onTouchStart={[Function]}
    to="http://www.topman.com"
  />
  <div
    className="MegaNav-subNavWrapper MegaNav-subNavWrapper--isNotTouch"
  >
    <SubNav
      category={
        Object {
          "categoryId": "12345",
          "columns": Array [
            Object {
              "span": "0",
              "subcategories": Object {
                "0": Object {
                  "newIn": Array [
                    Object {
                      "image": Object {
                        "span": "footer",
                      },
                    },
                    Object {
                      "label": "something to buy",
                    },
                  ],
                },
              },
            },
            Object {
              "span": "1",
              "subcategories": Object {
                "0": Object {
                  "0": Array [
                    Object {
                      "image": Object {
                        "span": 1,
                      },
                    },
                    Object {
                      "label": "red jeans",
                    },
                  ],
                  "jeans": Array [
                    Object {
                      "image": Object {
                        "span": "footer",
                      },
                    },
                    Object {
                      "label": "blue jeans",
                    },
                  ],
                },
              },
            },
            Object {
              "span": "2",
              "subcategories": Object {
                "0": Object {
                  "jeans": Array [
                    Object {
                      "image": Object {
                        "url": "",
                      },
                    },
                  ],
                  "socks": Array [
                    Object {
                      "label": "winter socks",
                    },
                  ],
                },
              },
            },
          ],
          "redirectionUrl": "/en/tsuk/category/new-in-this-week-2169932/new-in-fashion-6367514/N-8d7Zdgl?No=0&Nrpp=20&siteId=%2F12556",
          "url": "http://www.topman.com",
        }
      }
      isActive={false}
      key="12345"
      unselectCategory={[MockFunction]}
    />
  </div>
</li>
`;
